File: codigo/vigas/Euler-Bernoulli/viga_con_rotulas/K_EB_empotrado_rodillo_2.m
% Programa para deducir la matriz de rigidez de un elemento de viga de
% Euler-Bernoulli a partir de la solucion de la ecuacion diferencial
% que tiene una rotula a la derecha

clear, clc
syms x L V(x) M(x) t(x) v(x) EI EA

%% Se calcula la matrix de rigidez para la rotula en el lado derecho
K = sym(zeros(3));
for i = 1:3
    sol = dsolve(...       
           diff(V,x) == 0,    ... % se definen las ecuaciones
           diff(M,x) == V,    ... % diferenciales
           diff(t,x) == M/EI, ... 
           diff(v,x) == t,    ...
           v(0) == (i==1),    ... % con sus respectivas 
           t(0) == (i==2),    ... % condiciones de frontera
           v(L) == (i==3),    ...           
           M(L) == 0);

    K(:,i) = [ +subs(sol.V, x, 0)    % Y1  se evaluan las reacciones
               -subs(sol.M, x, 0)    % M1  verticales y los momentos
               -subs(sol.V, x, L) ]; % Y2  en los apoyos
end
K

%% Se calcula la matriz de rigidez para la rotula en el lado izquierdo
K = sym(zeros(3));
for i = 1:3
    sol = dsolve(...       
           diff(V,x) == 0,    ... % se definen las ecuaciones 
           diff(M,x) == V,    ... % diferenciales
           diff(t,x) == M/EI, ... 
           diff(v,x) == t,    ...
           v(0) == (i==1),    ... % con sus respectivas 
           M(0) == 0,         ... % condiciones de frontera  
           v(L) == (i==2),    ...           
           t(L) == (i==3));

    K(:,i) = [ +subs(sol.V, x, 0)    % Y1  se evaluan las reacciones
               -subs(sol.V, x, L)    % Y2  verticales y los momentos
               +subs(sol.M, x, L) ]; % M2  en los apoyos
end
K

File: codigo/vigas/Euler-Bernoulli_python/viga_con_rotulas/K_EB_empotrado_rodillo_2.py
# -*- coding: utf-8 -*-
'''
Programa para deducir la matriz de rigidez de un elemento de viga de
Euler-Bernoulli a partir de la solucion de la ecuacion diferencial
que tiene una rótula a la derecha.
'''

# %%Importación de librerías
import sympy as sp

# %%Definción de variables
x, L, V, M, t, w, EI = sp.symbols('x L  V M t w EI')
C1, C2, C3, C4 = sp.symbols('C1, C2, C3, C4')  # Constantes de integración

q = 0

# %%Se calcula la matriz de rigidez para la rotula en el lado derecho
K = sp.zeros(3)

V = sp.integrate(q, x) + C1  # Ecuaciónes calculadas con constantes.
M = sp.integrate(V, x) + C2
t = (sp.integrate(M, x) + C3)/EI
w = sp.integrate(t, x) + C4

for i in range(3):

    sol = sp.solve([w.subs(x, 0) - int((i == 0)),  # Condiciones de frontera
                    t.subs(x, 0) - int((i == 1)),
                    w.subs(x, L) - int((i == 2)),
                    M.subs(x, L)],
                   [C1, C2, C3, C4])

    constantes = [(C1, sol[C1]), (C2, sol[C2]), (C3, sol[C3]), (C4, sol[C4])]

    K[:, i] = [+ (V.subs(constantes)).subs(x, 0),  # Y1 se evaluan las
               - (M.subs(constantes)).subs(x, 0),  # M1 reacciones vert y los
               - (V.subs(constantes)).subs(x, L)]  # Y2 momentos en los apoyos

# Se imprime la solución
print('La matriz de rigidez para una rótula en el lado derecho es:')
print(f'{sp.pretty(K)}')

# %%Se calcula la matriz de rigidez para la rotula en el lado izquierdo
K = sp.zeros(3)

V = sp.integrate(q, x) + C1  # Ecuaciónes calculadas con constantes.
M = sp.integrate(V, x) + C2
t = (sp.integrate(M, x) + C3)/EI
w = sp.integrate(t, x) + C4

for i in range(3):

    sol = sp.solve([w.subs(x, 0) - int((i == 0)),  # Condiciones de frontera
                    M.subs(x, 0),
                    w.subs(x, L) - int((i == 1)),
                    t.subs(x, L) - int((i == 2))],
                   [C1, C2, C3, C4])

    constantes = [(C1, sol[C1]), (C2, sol[C2]), (C3, sol[C3]), (C4, sol[C4])]

    K[:, i] = [+ (V.subs(constantes)).subs(x, 0),  # Y1 se evaluan las
               - (V.subs(constantes)).subs(x, L),  # Y2 reacciones vert y los
               + (M.subs(constantes)).subs(x, L)]  # M2 momentos en los apoyos

# Se imprime la solución
print('\nLa matriz de rigidez para una rótula en el lado izquierdo es:')
print(f'{sp.pretty(K)}')
